File: 6KYU/The Shell Game/solution.clj
(ns the-shell-game)

(defn find-the-ball [start swaps]
  (if (empty? swaps)
    start
    (let [x (ffirst swaps)
          y (second (first swaps))]
      (cond (= start x) (recur y (rest swaps))
            (= start y) (recur x (rest swaps))
            :else (recur start (rest swaps))))))
			
;; alternate solution

(ns the-shell-game)

(defn swap [v i1 i2]
  (assoc v i2 (v i1) i1 (v i2)))

(defn find-the-ball [starting-pos swaps]
  "Given the starting position and a list of swaps, find the final position"
  (if (empty? swaps)
    starting-pos
  (loop [cups (assoc (vec (repeat (inc (apply max (flatten swaps))) :#)) starting-pos :O)
         cur-swaps swaps]
    (if (empty? cur-swaps)
      (.indexOf cups :O)
      (recur (swap cups (ffirst cur-swaps) (second (first cur-swaps))) (rest cur-swaps))))))
	  
	  

File: 6KYU/The Shell Game/solution.py
def find_the_ball(start, swaps):
    if len(swaps) == 0:
        return start
    else:
        x = swaps[0][0]
        y = swaps[0][1]
        if x == start:
            return find_the_ball(y, swaps[1::])
        elif y == start:
            return find_the_ball(x, swaps[1::])
        else:
            return find_the_ball(start, swaps[1::])